### DISCUSSION TITLE
Umruestung OpenDTU Fusion v2.1 Board fuer onBattery

### DISCUSSION THREAD

**Discussion by stegbth** [hoylabs/OpenDTU-OnBattery#944] [Ideas]
Hallo,

zuerst mal danke an alle fuer dieses tolle Projekt und Zusammenarbeit, um eine gemeinsame Loesung zu erschaffen.

Seit kurzem habe ich ein Balkonkraftwerk werte dieses mittels eine OpenDTU auf einem Fusion v2.1 Board aus.
Nun wird aber relativ viel Strom in's Netz eingspeist, da ich den Strom zu anderen Zeiten benoetige als produziert.
Die OpenDTU habe ich bereits auf OpenDTU on Battery vom 23.3.2024 umgeflasht und den Shelly 3EM pro angebunden.
 Die URL musste ich dabei umbauen - hier koennte man in den Beispielen die URLs mit aufnehmen:
 http://shelly3empro.home/rpc/EM.GetStatus?id=0
 JsonPfad: a_act_power
 
Gibt es ausserdem eine Moeglichkeit das Fusion v2.1 Board um eine VE.direct und Canbus Schnittstelle zu erweitern?
 Denn auf GPIO 17 liegt dort bereits eine LED D2 https://github.com/markusdd/OpenDTUFusionDocs welche ja beim DTUoB Board dem Canbus zugeordnet ist.

vg
Thomas

**Comment by markusdd** [comment]
Es ist sehr lustig, dass du das fragst.

Ich bin der Entwickler der Platine, da kommt in der Tat etwas ;)

Sind gerade im Zulauf, frisch bestellt.

<img width="383" alt="Top" src="https://github.com/helgeerbe/OpenDTU-OnBattery/assets/25175069/4cbafeb9-c4c4-47f1-8832-1a0a97454358">
<img width="374" alt="Bottom" src="https://github.com/helgeerbe/OpenDTU-OnBattery/assets/25175069/967295ee-cd66-439f-83a8-9162cd6a8fa9">

**Reply by kloppy1984** [reply]
Sieht top aus, klein und kompakt

**Reply by stegbth** [reply]
Hi @markusdd,
das sieht ja schon ziemlich genau nach dem aus, was sich suche ;)
Ich dachte noch an JST- und RJ45-Stecker, aber das kann man loesen, indem man doch die Kabel abzwickt.

Diese Platine wird dann einfach auf die Pfostenstecker mit den Mosi und 3,3V (SPI Displayheader)  gesteckt und hat im Gehäuse Platz?
In den Deckel muss man dann halt noch ein Loch bohren, um die Kabel einzuführen, oder?

Weiters braucht man noch ein aktualisiertes Config.json? Fuer das andere Pin Layout, richtig?

vg
Thomas

**Reply by markusdd** [reply]
Richtig. Sieht dann so aus:
![PXL_20240505_084754148 PORTRAIT](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/25175069/dd2f5980-994b-4570-894f-1633716fd778)

An der Doku arbeite ich gerade für das pin mapping die geht in kürze online

**Reply by stegbth** [reply]
Die Variante ist dann aber nur fuer einen VE.direct, oder?
 Was aber vermutlich fuer 90% der Faelle ausreicht.

**Reply by markusdd** [reply]
das ist eines der v1 boards vom prototyping, da gibt es nur 18 stück von.
Die gehen auch in den verkauf, haben aber nur support für einen victron.

Das richtige ist dann das v2: https://github.com/helgeerbe/OpenDTU-OnBattery/discussions/955

**Reply by markusdd** [reply]
Dokumentation ist jetzt online: https://github.com/markusdd/OpenDTUFusionDocs/blob/main/CANIso.md

**Reply by markusdd** [reply]
FYI, die v1 sind jetzt bestellbar: https://shop.allianceapps.io/products/opendtu-fusion-can-iso-shield-beta

**Reply by schlimmchen** [reply]
@markusdd Ich kann den Warenkorb nicht auschecken, da meine Adresse beanstandet wird. Details siehe Discord PM.

**Reply by markusdd** [reply]
@cat24max ist informiert, kleines issue in der shopeinstellung, einfach später heute nomma testen

**Reply by cat24max** [reply]
@schlimmchen Muesste jetzt gehen, sorry.

**Reply by schlimmchen** [reply]
@cat24max @markusdd Hat geklappt, danke!

**Comment by kloppy1984** [comment]
Ich hatte für den Flo eine Zusatzplatine entwickelt, für 1 MPPT + 1 Pylontech mit RJ45 Buchse
Angeschlossen sind per Flachkabel an die Display-Header.
das ganze einfach neben dran gelegt

Wenn Bedarf besteht, melden

**Reply by stegbth** [reply]
Hi @kloppy1984,
da gibt's dann bereits zwei Varianten? ;)
Hast Du evtl. noch ein Bild davon?
 Die Pinlayout ist dann ja identisch?
viele Gruesse

**Comment by typxxi** [comment]
An der Stelle die Frage, ob es da auch einen JK BMS Support geben wird oder gibt, denn das ist ja RS485 und nicht CAN BUS noch VE.
Hier sollen ein Victron und ein JK BMS am HM-800 / HM-1200 zum Einsatz kommen (Shelly Pro 3 EM im Verteiler).

Gibt es das Passendes für DT FUSION oder wird es eine Weiterentwicklung für eine Fusion on Battery Variant geben ?

Macht ja Sinn, gerade bei der Frage SPI, AC DC laden etc. 
DANKE

**Reply by schlimmchen** [reply]
Ich verstehe die Frage nicht so recht. Das OpenDTU FUSION hat ja einen RS485 Transceiver schon drauf. Der lässt sich auch ausdrücklich auswählen bei der Konfiguration der Batterie-Schnittstelle: "RS-485 Transceiver an der MCU". Der entsprechende Teil des Pin-Mapping sieht so aus:
```json
{
        "battery": {
            "rx": 16,
            "rxen": 15,
            "tx": 45,
            "txen": 46
        }
}
```
Bei mir läuft das auch genau so sehr gut.

Beantwortet das deine Frage bzw. räumt die Unsicherheiten aus?

**Reply by markusdd** [reply]
Danke, genau so isses.

**Reply by typxxi** [reply]
Der entsprechende Teil des Pin-Mapping sieht so aus:
> 
> ```json
> {
>         "battery": {
>             "rx": 16,
>             "rxen": 15,
>             "tx": 45,
>             "txen": 46
>         }
> }
> ```> 
> Bei mir läuft das auch genau so sehr gut.> 
> Beantwortet das deine Frage bzw. räumt die Unsicherheiten aus?

Vielen Dank, das hab ich so nicht gewusst, sondern gedacht, dass ich auch noch einen Shield a la Canbus bräuchte.

Ich wusste nicht , wie das sein würde, weil ich noch keinen Hoymiles hatte und mich durch den Dschungel von opendtu und den Varianten wie opendtu on battery am Durchkämpfen war. Ich komm aus der Growatt Ecke und da ist fast alles anders, eine ganz andere Welt des Auslesens bzw. Problemen mit der Cloud, wobei ich von großen Hausanlagen a la 30 kWp spreche. Wir haben 8 Growatt WR im Moment aktive aus diversen Gründen , da wir u.a. diy LFP Batterie betreiben, was nur mit 4,6 kW Hybriden geht, weil die großen 10 kW und mehr nur noch Hochvolt unterstützen.

Ich hab ja ewig gebraucht um zu blicken, dass das Fusionboard fast alles auf einem Board schon vereint, was ansonsten per Kabel zusammengehalten wurde oder mit breadboard. Ich hatte immer gedacht, hier würde noch irgendwo ein ESP32 aufgesteckt, bis ich dann mal genauer hinsah und dieses SMC43 Modul oder wie das hieß und den ESP 32 ich meine U mit dem Antennenanschluss erkannte und dann erst blickte, was das ist. 
Warte jetzt auf die Lieferung der Panels, bis dahin sollte der WR dann da sein und gleichfalls der Akku. Ist halt alles DIY und sozusagen noch im Fluß.

DANKE EUCH BEIDEN

**Reply by typxxi** [reply]
> Transceiver

DANKE,
aber eine Frage ist mir eben noch eingefallen. Ich habe jetzt testweise 2x HM-600 im Zulauf, die bald ankommen sollten.
Die Module speisen via mppt in Richtung Akku und kann ich dort dann auch 2 HM-600 anbinden und beide Richtung Nulleinspeisung trimmen?
Dann kann ich mir alles andere sparen.

**Comment by akihilananda** [comment]
Funktioniert mit CAN/Iso Shield einwandfrei. Daten von victron mppt 100 20 und smart shunt werden in OpenDTUon battery  angezeigt.
Pinmapping:
victron:   rx 10
battery:  rx 11

https://github.com/markusdd/OpenDTUFusionDocs/blob/main/CANIso.md

**Comment by Hutzelbrot** [comment]
Ist der zusätzliche Anschluß des Huawei Netzteils an OpenDTU Fusion mit CAN/Iso Shield möglich?
Meine Konfiguration: NRF24, 2xVictron MPPT und Pytes Batterie.

**Reply by markusdd** [reply]
nach meinem verständnis ist dafür ein weiteres SPI-basiertes CAN Board erforderlich, an sich, solange kein PoE-Shield im EInsatz ist, könnte man probieren über entsprechendes Pin-mapping dies z.B. an die JTAG/Uart-header anzuschließen.

Ist allerdings definitiv ein exotisches Setup und ich weiß nicht wo der software support diesbezüglich steht.

**Reply by schlimmchen** [reply]
> Ist allerdings definitiv ein exotisches Setup

Hm, finde ich nicht.

> ich weiß nicht wo der software support diesbezüglich steht.

Wie von Markus erwähnt brauchst du ein zusätzliches CAN Board. Das wird per SPI angebunden, und den SPI Bus für den CMT2003 musst du dann aufgeben und umlenken auf Pins, an die du noch herankommst (JTAG header). Da du mit NRF24 unterwegs bist, sollte das gehen.

**Reply by markusdd** [reply]
Na wenn der Chef sagt es ist nicht exotisch passt es ja haha.

Aber ja, genauso wäre es. Was ist eigentlich der Grund warum da ein normaler CAN transceiver nicht ausreicht?

Weil ansonsten könnte man die Batterie ja auch auf rs485 schicken und den Huawei an den CAN

**Reply by schlimmchen** [reply]
> Was ist eigentlich der Grund warum da ein normaler CAN transceiver nicht ausreicht?

Was meinst du denn mit normal? Es ist jedenfalls so, dass der CAN Controller für diesen Huawei Charger bisher nur per SPI angesprochen werden kann (Software-Thema) und der SPI-Bus eine andere Frequenz hat als der für den CMT2003. Daher schließen die sich aus. Und dann sind zu wenig SPI Busse da, korrekt? Puh, gefährliches Halbwissen gerade bei mir, muss ich zugeben...

> Weil ansonsten könnte man die Batterie ja auch auf rs485 schicken und den Huawei an den CAN

Das wurde auch schon zurecht vorgeschlagen. Müsste mal jemand™ implementieren.

**Reply by markusdd** [reply]
Naja wir haben ja nun bei Pylontech und Pytes direkte Can kommunikation.
Warum also braucht der Huawei nun nen extra board? Weil das jemand rumliegen hatte und deswegen alles für SPI statt direkt CAN geschrieben hat? Mir ist halt nicht klar was dieses extra board können soll was der CAN-Transceiver nicht auch direkt senden könnte.

Und am Ende ist CAN natürlich eh ein Multi-Bus, also an sich könnte der eine CAN-Port ja auch mehrere devices ansprechen.

**Reply by schlimmchen** [reply]
Den Gedanken hatte ich noch nicht, danke dafür. Wie sieht das aus mit unterschiedlichen Busgeschwindigkeiten? Wenn das Huawei x Baud haben möchte, die Pytes/Pylontechs aber y Baud, dann gibt's ein Problem?

Ansonsten:

> dass der CAN Controller für diesen Huawei Charger bisher nur per SPI angesprochen werden kann (Software-Thema)

Mit Betonung auf "Software-Thema". Das ganze Ding (Steuerung für das Huawei Netzteil) ist nuneinmal so implementiert, dass es einen (bestimmten) CAN-Transceiver über SPI ansprechen möchte...

**Reply by markusdd** [reply]
angesprochene Geräte sollten eigentlich eine spanne von Baudraten unterstützen. In Autos hängen ja potenziell sehr viele verschiedene Geräte an sowas, das muss ja auch gehen.

aber ja, dann ist dieses Huawei-Ding vermutlich wirklich eine reine Softwaresache die der ursprüngliche Autor so gemacht hat.
Es macht halt garkeinen Sinn dafür einen SPI zu verballern, was sonst auch mit 2 pins geht.

Wenn es nicht einen sehr speziellen Grund gibt verstehe ich es rein technisch jedenfalls nicht.

**Reply by schlimmchen** [reply]
Ich stimme dir zu.

> In Autos hängen ja potenziell sehr viele verschiedene Geräte an sowas, das muss ja auch gehen.

Also ich hab dazu im Studium gelernt (>10a), dass es auch deshalb mehrere CAN-Busse in Autos gibt, sowohl weil die einen wichtiger sind als die anderen (Entertainment nicht mit drive train mischen), aber auch, weil die dann auf verschiedenen Baud laufen, weil die unterschiedlich viel Durchsatz brauchen.

ChatGPT stimmt mir zu und erinnert mich, dass insb. das Arbitrierungsverfahren in CAN nicht funktionieren kann, wenn es unterschiedliche Baudraten gibt. Ganz kurz: Es können mehrere Teilnehmer gleichzeitig versuchen eine Nachricht auf den Bus zu legen, und es gewinnt dann derjenige mit der ersten Null in der Adresse, d.h. die machen media sensing während sie die Adresse schreiben, was nur klappt wenn alle den gleichen Takt haben.

Also das wäre erst noch zu prüfen, ob die auf den gleichen Bus könnten. Und wenn dem so ist, muss die Software aufgehübscht werden. Und sowieso könnte man eigentlich einen besseren Transceiver nehmen, sodass kein SPI verballert wird.

**Reply by skippermeister** [reply]
Soweit ich den CAN Bus verstanden habe, müssen alle Geräte auf einem CAN Bus mit der gleichen Taktrate arbeiten. Somit benötigen wir für das HUAWEI Netzteil einen CAN Bus und für die über CAN Bus angeschlossene Pylontech oder Pytes Batterie einen weiteren CAN Bus. Und da der ESP32 nur einen CAN Bus hat hat wird ein separater CAN Controller benötigt. 

Es gibt aber auch ein I2C to MCP2515 Modul von Seeed Studio. Das Modul wird über I2C angeschlossen, benötigt 2 GPIO2.
Das Modul hat einen ATMEGA168P, den MCP2515 CAN Bus Controller und den CAN Bus Treiber. 
Das Modul wird per Polling-Verfahren abgefragt.
Die Implementierung sollte relativ simple sein und ersetzt dann den über SPI angeschlossen MCP2515 CAN Bus Controller.
Ich habe mir so ein Modul mal gekauft, bin aber bislang nicht dazu gekommen, es auszuprobieren. 

Eine andere Möglichkeit wäre das von mir entwickelte PylontechRS485Receiver Modul einzubinden (siehe mein Fork).  Das müsste für die Pytes Batterien noch portiert werden. Sowohl Pylontech und Pytes können per RS485 angesprochen werden.

**Comment by Harisjems** [comment]
Hallo zusammen, und frohes Neues allen zusammen.

Ich versuche verzweifelt seit gestern, das Waveshare E-Paper Display 2.13" auf dem Fusion Board V2 zum Laufen zu bekommen, ohne Erfolg. Es scheitert am Pin-Mapping und meinem Verständnis.
OpenDTU on battery, aktuelle Entwicklerversion.

Ist dieses Anschlussschema überhaupt korrekt?

"display": {
"type": 6,
"data": 9,
"clk": 10,
"cs": 11,
"reset": 13,
"busy": 14,
"dc": 12
}

MISO, MOSI, BUSY Anschlüsse bez. an der Fusion Board: An welche Pins am Display sollen diese angeschlossen werden? Ich vermute, dass ich da Fehler gemacht habe.

**Reply by markusdd** [reply]
Hi ebenso ein Frohes Neues!

Ich kenne das Display leider nicht persönlich, weiß aber dass Leute das benutzen am Fusion.
Bist du schonmal auf dem Community Discord gewesen? Da gibt es sicher jemand der dir weiterhelfen könnte mit einer funktionierenden Pinconfig.

**Comment by Harisjems** [comment]
Grüß dich Markus,

Mir würde schon helfen, wenn ich die Reihenfolge der Pinbelegung auf dem Fusion Board V2.3 richtig lokalisieren/deuten könnte :-)

Laut Datenblatt und SPI-Schnittstelle ist der erste GPIO-Pin von oben nach unten gezählt GPIO9 (MOSI), der zweite GPIO10 (SCK) und die letzten beiden GPIO13 (RST), GPIO14 (MISO).

Korrekt!

**Reply by markusdd** [reply]
Was fehlt dir denn zur 'Deutung' außer dieser übersicht^^? 
![grafik](https://github.com/user-attachments/assets/37f7d80d-5895-4160-86fb-35584762870b)
